web/src/lib/actions/input.ts

```diff
@@ -0,0 +1,229 @@
+  ...RESERVED_SHORTCUTS.win,
+  ...RESERVED_SHORTCUTS.linux,
+  ...RESERVED_SHORTCUTS.browser,
+];
+
+const attachmentFactory = (listenerFactory: KeyDownListenerFactory) => (element: KeyTargets) => {
+  const listener = listenerFactory(element);
+  element.addEventListener('keydown', listener as EventListener);
+  return () => {
+    element.removeEventListener('keydown', listener as EventListener);
+  };
+};
+
+function isKeyboardEvent(event: Event): event is KeyboardEvent {
+  return 'key' in event;
+}
+
+/** Determines whether an event should be ignored. The event will be ignored if:
+ *  - The element dispatching the event is not the same as the element which the event listener is attached to
+ *  - The element dispatching the event is an input field
+ */
+export const shouldIgnoreEvent = (event: KeyboardEvent | ClipboardEvent): boolean => {
+  if (event.target === event.currentTarget) {
+    return false;
+  }
+  const type = (event.target as HTMLInputElement).type;
+  return ['textarea', 'text', 'date', 'datetime-local', 'email', 'password'].includes(type);
+};
+
+function isPromise<T = unknown>(obj: unknown): obj is Promise<T> {
+  return (
+    obj !== null &&
+    (typeof obj === 'object' || typeof obj === 'function') &&
+    typeof (obj as { then?: unknown }).then === 'function'
+  );
+}
+
+function checkForReservedShortcuts(
+  shortcuts: KeyCombo[],
+  handler: (event: KeyboardEvent) => void,
+): (event: KeyboardEvent) => void {
+  const formatCombo = (combo: KeyCombo) => {
+    const parts = [];
+    if (combo.ctrl) {
+      parts.push('Ctrl');
+    }
+    if (combo.alt) {
+      parts.push('Alt');
+    }
+    if (combo.shift) {
+      parts.push('Shift');
+    }
+    if (combo.meta) {
+      parts.push('Meta');
+    }
+    parts.push(combo.key);
+    return parts.join('+');
+  };
+
+  for (const shortcut of shortcuts) {
+    for (const reserved of ALL_RESERVED_SHORTCUTS) {
+      // Check if shortcuts match (comparing all properties)
+      if (
+        shortcut.key.toLowerCase() === reserved.key.toLowerCase() &&
+        !!shortcut.ctrl === !!reserved.ctrl &&
+        !!shortcut.alt === !!reserved.alt &&
+        !!shortcut.shift === !!reserved.shift &&
+        !!shortcut.meta === !!reserved.meta
+      ) {
+        console.error(
+          `[Keyboard Shortcut Warning] Attempting to register reserved shortcut: ${formatCombo(shortcut)}. ` +
+            `This shortcut is reserved by the OS or browser and may not work as expected.`,
+        );
+        return () => void 0;
+      }
+    }
+  }
+  return handler;
+}
+
+export const keyDownListenerFactory = (
+  isActiveFactory: () => () => boolean,
+  options: { ignoreInputFields?: boolean },
+  shortcuts: KeyCombo[],
+  callback: ShortcutCallback,
+) =>
+  checkForReservedShortcuts(shortcuts, (event: KeyboardEvent) => {
+    const isActive = isActiveFactory();
+    if (!isActive() || !isKeyboardEvent(event) || ((options.ignoreInputFields ?? true) && shouldIgnoreEvent(event))) {
```

Comment:
Nit: default `options.ignoreInputFields` in the parameter list and remove the fallback here